internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -640,19 +640,52 @@ func (r *ClusterDeploymentsReconciler) updateIfNeeded(ctx context.Context,
 	updateString(spec.BaseDomain, cluster.BaseDNSDomain, &params.BaseDNSDomain)
 
 	if len(clusterInstall.Spec.Networking.ClusterNetwork) > 0 {
-		updateString(clusterInstall.Spec.Networking.ClusterNetwork[0].CIDR, cluster.ClusterNetworkCidr, &params.ClusterNetworkCidr)
-		hostPrefix := int64(clusterInstall.Spec.Networking.ClusterNetwork[0].HostPrefix)
-		if hostPrefix > 0 && hostPrefix != cluster.ClusterNetworkHostPrefix {
-			params.ClusterNetworkHostPrefix = swag.Int64(hostPrefix)
+		newClusterNetworks := clusterNetworksEntriesToArray(clusterInstall.Spec.Networking.ClusterNetwork)
+		if len(newClusterNetworks) != len(cluster.ClusterNetworks) {
+			params.ClusterNetworks = newClusterNetworks
 			update = true
+		} else {
+			for index := range newClusterNetworks {
+				if newClusterNetworks[index].Cidr != cluster.ClusterNetworks[index].Cidr ||
+					newClusterNetworks[index].HostPrefix != cluster.ClusterNetworks[index].HostPrefix {
+					params.ClusterNetworks = newClusterNetworks
+					update = true
+					break
+				}
+			}
 		}
 	}
 	if len(clusterInstall.Spec.Networking.ServiceNetwork) > 0 {
-		updateString(clusterInstall.Spec.Networking.ServiceNetwork[0], cluster.ServiceNetworkCidr, &params.ServiceNetworkCidr)
+		newServiceNetworks := serviceNetworksEntriesToArray(clusterInstall.Spec.Networking.ServiceNetwork)
+		if len(newServiceNetworks) != len(cluster.ServiceNetworks) {
+			params.ServiceNetworks = newServiceNetworks
+			update = true
+		} else {
+			for index := range newServiceNetworks {
+				if newServiceNetworks[index].Cidr != cluster.ServiceNetworks[index].Cidr {
+					params.ServiceNetworks = newServiceNetworks
+					update = true
+					break
+				}
+			}
+		}
 	}
 	if len(clusterInstall.Spec.Networking.MachineNetwork) > 0 {
-		updateString(clusterInstall.Spec.Networking.MachineNetwork[0].CIDR, cluster.MachineNetworkCidr, &params.MachineNetworkCidr)
+		newMachineNetworks := machineNetworksEntriesToArray(clusterInstall.Spec.Networking.MachineNetwork)
+		if len(newMachineNetworks) != len(cluster.MachineNetworks) {
+			params.MachineNetworks = newMachineNetworks
+			update = true
+		} else {
+			for index := range newMachineNetworks {
+				if newMachineNetworks[index].Cidr != cluster.MachineNetworks[index].Cidr {
+					params.MachineNetworks = newMachineNetworks
+					update = true
+					break
+				}
+			}
```

Comment:
Seems repetitive (for each network type), see if you can extract the logic to a function and reuse it